integration/secret_cert_test.go

```diff
@@ -127,3 +129,130 @@ func TestCertificateUpload(t *testing.T) {
 
 	testenv.Test(t, f)
 }
+
+func TestCertificateUploadForOAS(t *testing.T) {
+	var (
+		tlsSecretName = "test-oas-cert-upload-secret"
+
+		testNS string
+		tykOas *v1alpha1.TykOasApiDefinition
+		tykEnv environment.Env
+		tykCtx context.Context
+	)
+	f := features.New("test certificate upload").
+		Setup(func(ctx context.Context, t *testing.T, envConf *envconf.Config) context.Context {
+			eval := is.New(t)
+			var err error
+
+			testNS, _ = ctx.Value(ctxNSKey).(string) // nolint: errcheck
+
+			// Obtain Environment configuration to be able to connect Tyk.
+			tykEnv, err = generateEnvConfig(ctx, envConf)
+			eval.NoErr(err)
+
+			res := semver.Compare(tykEnv.TykVersion, "v4.1")
+			if res < 0 {
+				t.Skip("OAS support is added in Tyk in v4.1")
+			}
+
+			tykCtx = tykClient.SetContext(context.Background(), tykClient.Context{
+				Env: tykEnv,
+				Log: log.NullLogger{},
+			})
+
+			// Create TLS secret
+			_, err = createTestTlsSecret(ctx, testNS, envConf, func(s *v1.Secret) {
+				s.Name = tlsSecretName
+			})
+			eval.NoErr(err)
+
+			tykOASDoc := `{
+				"info": {
+				  "title": "Petstore",
+				  "version": "1.0.0"
+				},
+				"openapi": "3.0.3",
+				"components": {},
+				"paths": {},
+				"x-tyk-api-gateway": {
+				  "info": {
+					"name": "Petstore",
+					"state": {
+					  "active": true
+					}
+				  },
+				  "upstream": {
+					"url": "https://petstore.swagger.io/v2"
+				  },
+				  "server": {
+					"listenPath": {
+					  "value": "/petstore/",
+					  "strip": true
+					},
+					"clientCertificates":{
+					  "enabled": true
+					}
+				  }
+				}
+			  }`
+
+			// Create test api definition
+			tykOas, _, err = createTestOASApi(ctx, testNS, envConf, tykOASDoc,
+				func(oas *v1alpha1.TykOasApiDefinition) *v1alpha1.TykOasApiDefinition {
+					oas.Spec.ClientCertificateRef = []string{tlsSecretName}
+
+					return oas
+				})
+			eval.NoErr(err)
+
+			err = waitForTykResourceCreation(envConf, tykOas)
+			eval.NoErr(err)
+
+			return ctx
+		}).Assess("validate certificate is uploaded",
+		func(ctx context.Context, t *testing.T, envConf *envconf.Config) context.Context {
+			eval := is.New(t)
+
+			// Validate the certificate is uploaded
+			err := wait.For(func() (done bool, err error) {
+				tykOASApi, err := klient.Universal.TykOAS().Get(tykCtx, tykOas.Status.ID)
+				if err != nil {
+					t.Log("Failed to fetch Tyk OAS API", "id", tykOas.Status.ID, "error", err)
+
+					return false, err
```

Comment:
Returning an err will halt the execution. So, sometimes - especially in OSS gateway - creation of files takes a while. I'd suggest retry getting the OAS until timeout occurs